src/js/api/posts/create.mjs

```diff
@@ -0,0 +1,27 @@
+import { BASE_URL, optionsWithBody } from "../index.mjs";
+import {
+  renderSingleProfile,
+  renderUserSpecificPosts,
+} from "../../renders/index.mjs";
+import { errorMessage } from "../../templates/errorMessage.mjs";
+
+const createPostURL = `${BASE_URL}/posts`;
+
+/**
+ * Send the data to create a new post.
+ * Re-render the page after sending the data.
+ *
+ * @param {object} postData data input by the user
+ */
+export async function createPost(postData) {
+  try {
+    const options = optionsWithBody("POST", postData);
+    const response = await fetch(createPostURL, options);
+    const result = await response.json();
+    console.log(result);
+    renderUserSpecificPosts();
+  } catch (e) {
+    console.log(e);
+    errorMessage();
```

Comment:
Wait, If this is the case, what should API data function do with the error it catches in try catch? Or you mean remove try catch from this function?

Reply:
This is why I don't use try/catch in API functions. Their failure is important information and a catch block `suppresses failure`.

My perfect API function:

```js
import { DEFAULT_OPTIONS } from "defaults";

export const update = async (url, payload, options) => {
  const response = await fetch(url, { ...DEFAULT_OPTIONS, ...options, body: JSON.stringify(payload) });
  const data = await response.json();
  if (response.ok) {
    return data;
  }
  
  throw new Error(JSON.stringify(data));
}
```

If it fails, you will know about it!

Reply:
This function can either:
1. Succeed and return JSON
2. Fail with a detailed error message string

You would `consume` it inside a narrative function:

```js
import api from "api";

export async function onUpdateFormSubmit({ target: form }) {
  const payload = Object.fromEntries((new FormData(form)).entries());
  try {
    await api.update(form.action, payload);
  } catch(error) {
    const message = JSON.parse(error);
    console.warn(message, error);
    alert("Update failed!")
  }
}
```

```js
import { onUpdateFormSubmit } from "events";

async function setupUpdatePage() {
  const form = document.querySelector('form');
  form.addEventListener(onUpdateFormSubmit);
}
```

Reply:
1. It would probably just be the Auth header, and content type. The idea with using `...` is to merge objects together, so if you want to completely override the defaults you can. This keeps it flexible.
2. That is an example of `object destructuring` and `reassignment`:

```js
const example = { title: "Hello world" };
const { title } = example;
const { title: testString } = example;
console.log(title === testString);
```